src/strategies/strategy/rl-dqn/tradeAgent.ts

```diff
@@ -0,0 +1,202 @@
+    this.reset();
+  }
+
+  reset() {
+    this.cumulativeReward_ = 0;
+    this.env.reset();
+  }
+
+  /**
+   * Play one step of the game.
+   *
+   * @returns {number | null} If this step leads to the end of the game,
+   *   the total reward from the game as a plain number. Else, `null`.
+   */
+  playStep() {
+    this.epsilon =
+      this.frameCount >= this.epsilonDecayFrames
+        ? this.epsilonFinal
+        : this.epsilonInit + this.epsilonIncrement_ * this.frameCount;
+    this.frameCount++;
+
+    // The epsilon-greedy algorithm.
+    let action;
+    const state = this.env.getState().flat();
+    if (Math.random() < this.epsilon) {
+      // Pick an action at random.
+      const actionNum = Math.random();
+      if (actionNum < 0.25) {
+        action = 0;
+      }
+      if (actionNum >= 0.25 && actionNum < 0.5) {
+        action = 1;
+      }
+      if (actionNum >= 0.5 && actionNum < 0.75) {
+        action = 2;
+      }
+      if (actionNum >= 0.75) {
+        action = 3;
+      }
+    } else {
+      //   Greedily pick an action based on online DQN output.
+      tf.tidy(() => {
+        // const flatState = state.flat(); // Convert state array to a flat array
+        const prediction = this.onlineNetwork.predict(
+          tf.tensor1d(state).reshape([1, -1]),
+        ) as tf.Tensor<tf.Rank>;
+        const actionIndex = prediction.argMax(-1).dataSync()[0];
+        action = this.ALL_ACTIONS[actionIndex];
+        // console.log(action);
+      });
+    }
+    const nextState = this.env.step(action);
+    this.replayMemory.append([
+      state,
+      action,
+      nextState.reward,
+      Object.values(nextState).slice(0, 5),
+      this.env.done,
+    ]);
+
+    this.cumulativeReward_ += nextState.reward;
+    const output = {
+      action,
+      cumulativeReward: this.cumulativeReward_,
+      done: this.env.done,
+    };
+    // if (done) {
+    //   this.reset();
+    // }
+    return output;
+  }
+
+  /**
+   * Perform training on a randomly sampled batch from the replay buffer.
+   *
+   * @param {number} batchSize Batch size.
+   * @param {number} gamma Reward discount rate. Must be >= 0 and <= 1.
+   * @param {tf.train.Optimizer} optimizer The optimizer object used to update
+   *   the weights of the online network.
+   */
+  trainOnReplayBatch(batchSize, gamma, optimizer) {
+    // Get a batch of examples from the replay buffer.
+    const batch = this.replayMemory.sample(batchSize);
+    const lossFunction = () =>
+      tf.tidy(() => {
+        // console.log(batch.map((example) => example[0]));
+        const stateTensor = tf.tensor2d(batch.map((example) => example[0]));
+        const actionTensor = tf.tensor1d(
+          batch.map((example) => example[1]),
+          'int32',
+        );
+        let qs = this.onlineNetwork.apply(stateTensor, {
+          training: true,
+        });
+        qs = (qs as tf.Tensor<tf.Rank>)
+          .mul(tf.oneHot(actionTensor, this.NUM_ACTIONS))
```

Comment:
- logical operators cannot be used in function parameters

Reply:
extract the logical operators